### PR TITLE
dc2 weak lensing

### PR DESCRIPTION
**Pull request by shreyasc30**

### CODE REVIEW

**Comment by shreyasc30** [comment]
waiting on #1047 for image normalizers & potential dc2 changes

File: bliss/cached_dataset.py [hunk 1]
```diff
@@ -149,7 +149,10 @@ def __getitem__(self, index):
             self.buffered_file_index = converted_index
             with open(self.file_paths[converted_index], "rb") as f:
                 self.buffered_data = torch.load(f)
-        output_data = self.buffered_data[converted_sub_index]
+        try:
+            output_data = self.buffered_data[converted_sub_index]
+        except KeyError:
```
Comment (jeff-regier) [inline bliss/cached_dataset.py:154]: Why are we getting key errors here?

File: bliss/catalog.py [hunk 2]
```diff
@@ -190,27 +196,44 @@ def to_full_catalog(self):
...
             params[param_name] = param
 
-        params["n_sources"] = reduce(self["n_sources"], "b nth ntw -> b", "sum")
+        if self.locations_sources:
+            params["n_sources"] = reduce(self["n_sources"], "b nth ntw -> b", "sum")
+
+        if "shear" in self:  # pylint: disable=too-many-branches
+            params["shear"] = self.get("shear")
+        if "convergence" in self:  # pylint: disable=too-many-branches
+            params["convergence"] = self.get("convergence")
```
Comment (jeff-regier) [inline bliss/catalog.py:236]: I'm reluctant to add special cases to this function that will make it even more complicated; it's so complex already. Is there a way to avoid these changes?

File: bliss/catalog.py [hunk 3]
```diff
@@ -116,7 +116,13 @@ class TileCatalog(BaseTileCatalog):
     galaxy_params_index = {k: i for i, k in enumerate(galaxy_params)}
 
     def __init__(self, tile_slen: int, d: Dict[str, Tensor]):
-        self.max_sources = d["locs"].shape[3]
+        self.locations_sources = True
+
+        if "locs" in d:
+            self.max_sources = d.get("locs").shape[3]
```
Comment (jeff-regier) [inline bliss/catalog.py:122]: Could you use `BaseTileCatalog` instead of adding special cases to `TileCatalog`?

File: bliss/conf/base_config.yaml [hunk 4]
```diff
@@ -239,8 +239,8 @@ surveys:
         prepare_data_processes_num: 4
         data_in_one_cached_file: 1250
         splits: 0:80/80:90/90:100
-        batch_size: 64
-        num_workers: 4
+        batch_size: 1
+        num_workers: 1
```
Comment (jeff-regier) [inline bliss/conf/base_config.yaml:243]: Better to make the changes in config file that extends base_config.yaml, rather than directly in base_config.yaml

File: bliss/conf/base_config.yaml [hunk 5]
```diff
@@ -303,7 +302,7 @@ train:
     matmul_precision: high
 
 predict:
-    dataset: ${surveys.sdss}
+    dataset: ${surveys.dc2}
```
Comment (jeff-regier) [inline bliss/conf/base_config.yaml:305]: same here

File: bliss/surveys/dc2.py [hunk 6]
```diff
@@ -320,7 +320,6 @@ def read_image_for_bands(self, image_index):
                 self.image_lim[0], self.image_lim[1]
             )
 
-            image += bg
             image_list.append(image)
             bg_list.append(bg)
```
Comment (jeff-regier) [inline bliss/surveys/dc2.py:324]: Perhaps we don't need to record `bg` at all any more?

File: case_studies/weak_lensing/lensing_dc2.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        dc2_cat_path: str,
+        image_slen: int,  # assume square images: image_slen x image_slen
+        tile_slen: int,
+        splits: str,
+        batch_size: int,
+        num_workers: int,
+        cached_data_path: str,
+        **kwargs,
+    ):
+        super().__init__(
```
Comment (jeff-regier) [inline case_studies/weak_lensing/lensing_dc2.py:23]: Can you pass `*args` and `**kwargs` to the parent class, instead of explicitly listing all of them?

File: case_studies/weak_lensing/lensing_encoder.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        var_dist: VariationalDist,
+        sample_image_renders: MetricCollection,
+        mode_metrics: MetricCollection,
+        sample_metrics: Optional[MetricCollection] = None,
+        optimizer_params: Optional[dict] = None,
+        scheduler_params: Optional[dict] = None,
+        reference_band: int = 2,
+        **kwargs,
+    ):
+        super().__init__(
```
Comment (jeff-regier) [inline case_studies/weak_lensing/lensing_encoder.py:31]: Same here, how about passing `*args` and `**kwargs` to the parent class? (and removing named arguments in the construct that are just passed on to `Encoder`)

File: case_studies/weak_lensing/lensing_encoder.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    def predict_step(self, batch, batch_idx, dataloader_idx=0):
+        """Pytorch lightning method."""
+        with torch.no_grad():
+            return {
+                "mode_cat": self.sample(batch, use_mode=True),
+                # we may want multiple samples
+                "sample_cat": self.sample(batch, use_mode=False),
+            }
+
+    def configure_optimizers(self):
```
Comment (jeff-regier) [inline case_studies/weak_lensing/lensing_encoder.py:168]: Could you omit some of these methods, which are the same in the parent class?

File: case_studies/weak_lensing/lensing_metrics.py [hunk 10]
```diff
@@ -6,35 +6,62 @@ class LensingMapMSE(Metric):
...
+        baseline_convergence_sq_err = ((true_convergence - baseline_pred_convergence) ** 2).sum()
+
+        self.shear1_sum_squared_err += shear1_sq_err
+        self.baseline_shear1_sum_squared_err += baseline_shear1_sq_err
+        self.shear2_sum_squared_err += shear2_sq_err
+        self.baseline_shear2_sum_squared_err += baseline_shear2_sq_err
+        self.convergence_sum_squared_err += convergence_sq_err
+        self.baseline_convergence_sum_squared_err += baseline_convergence_sq_err
+
+        self.total = torch.tensor(  # pylint: disable=attribute-defined-outside-init
```
Comment (jeff-regier) [inline case_studies/weak_lensing/lensing_metrics.py:48]: Would it work to set `self.total = None` in the constructor? (rather than disabling this check)

File: bliss/catalog.py (hunk 3 above)
Comment (shreyasc30) [reply bliss/catalog.py:122]: will take a look. The main issue here and in the previous case is that TileCatalog assumes ''locs'' is in the data, but we provide no such parameter

File: bliss/catalog.py (hunk 2 above)
Comment (shreyasc30) [reply bliss/catalog.py:236]: this change was made since shear and convergence don't follow the generalized pattern requiring indices_of_on_sources  that the catalog mandates. will take a look at the version after your new PR and see what can be done here.

File: case_studies/weak_lensing/lensing_dc2.py (hunk 7 above)
Comment (shreyasc30) [reply case_studies/weak_lensing/lensing_dc2.py:23]: they are listed since some of them are hardcoded as dummy values: i.e. min_flux_for_loss, max_sources_per_tile, image_lim, prepare_data_processes, data_in_one_cached_file, train_transforms, nontrain_transforms. 

I can instead add these hardcoded values to kwargs but this provides more visibility into the subclassing imo

File: case_studies/weak_lensing/lensing_encoder.py (hunk 8 above)
Comment (shreyasc30) [reply case_studies/weak_lensing/lensing_encoder.py:31]: will remove named arguments, but same issue as above with hardcoding certain params

File: bliss/cached_dataset.py (hunk 1 above)
Comment (shreyasc30) [reply bliss/cached_dataset.py:175]: investigating. was initially done as a workaround for file names, perhaps redundant now. will fix if so

File: bliss/catalog.py (hunk 2 above)
Comment (shreyasc30) [reply bliss/catalog.py]: resolved in refactor

File: bliss/catalog.py (hunk 3 above)
Comment (shreyasc30) [reply bliss/catalog.py]: resolved in refactor

File: bliss/conf/base_config.yaml (hunk 4 above)
Comment (shreyasc30) [reply bliss/conf/base_config.yaml]: resolved

File: bliss/conf/base_config.yaml (hunk 5 above)
Comment (shreyasc30) [reply bliss/conf/base_config.yaml]: resolved

File: bliss/surveys/dc2.py (hunk 6 above)
Comment (shreyasc30) [reply bliss/surveys/dc2.py:324]: waiting on incoming dc2 change (in the open yd PR) that will do the same

File: case_studies/weak_lensing/lensing_encoder.py (hunk 9 above)
Comment (shreyasc30) [reply case_studies/weak_lensing/lensing_encoder.py]: resolved

**Comment by shreyasc30** [comment]
refactored the tile catalog and made the small changes.
imo, the explicit stating of input parameters is better than just accepting **kwargs; that is there since the base_config provides arguments that can't be replaced (i.e. without it, there's some cases of "unexpected keyword ___ found" , but in this case we ensure that any missing arguments in lensing_config are noted quickly by the explicit constructor.

File: bliss/cached_dataset.py (hunk 1 above)
Comment (shreyasc30) [reply bliss/cached_dataset.py:175]: waiting for data gen to fix this, still outstanding -- will test & fix first thing tomorrow

File: case_studies/weak_lensing/lensing_encoder.py (hunk 8 above)
Comment (jeff-regier) [reply case_studies/weak_lensing/lensing_encoder.py:27]: Ok sounds good

**Review by jeff-regier — CHANGES_REQUESTED** [review]
Looking good. Please mark this PR as ready for review once the merge conflicts are resolved.

**Review by jeff-regier — APPROVED** [review]
